{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNPs data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>gene</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15</td>\n",
       "      <td>67138395</td>\n",
       "      <td>T&gt;A|C</td>\n",
       "      <td>snv</td>\n",
       "      <td>959504</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>SMAD3</td>\n",
       "      <td>T:0.034345:172:1000Genomes|T:0.000259:1:ALSPAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15</td>\n",
       "      <td>67066140</td>\n",
       "      <td>G&gt;A|C</td>\n",
       "      <td>snv</td>\n",
       "      <td>1061427</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>5_prime_UTR_variant;genic_upstream_transcript_...</td>\n",
       "      <td>SMAD3;LOC102723493</td>\n",
       "      <td>A:0.223642:1120:1000Genomes|A:0.233783:901:ALS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>67164997</td>\n",
       "      <td>A&gt;G|T</td>\n",
       "      <td>snv</td>\n",
       "      <td>1065080</td>\n",
       "      <td>likely-benign;benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>5_prime_UTR_variant;coding_sequence_variant;sy...</td>\n",
       "      <td>SMAD3</td>\n",
       "      <td>A:0.165935:831:1000Genomes|A:0.111572:430:ALSP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>67137885</td>\n",
       "      <td>T&gt;A|C|G</td>\n",
       "      <td>snv</td>\n",
       "      <td>1866318</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_varia...</td>\n",
       "      <td>SMAD3</td>\n",
       "      <td>T:0.034345:172:1000Genomes|T:0.000259:1:ALSPAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>67191959</td>\n",
       "      <td>C&gt;G|T</td>\n",
       "      <td>snv</td>\n",
       "      <td>2278670</td>\n",
       "      <td>benign</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>SMAD3</td>\n",
       "      <td>T:0.252796:1266:1000Genomes|T:0.228075:879:ALS...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #chr       pos variation variant_type   snp_id clinical_significance  \\\n",
       "0   15  67138395     T>A|C          snv   959504                benign   \n",
       "1   15  67066140     G>A|C          snv  1061427                benign   \n",
       "2   15  67164997     A>G|T          snv  1065080  likely-benign;benign   \n",
       "3   15  67137885   T>A|C|G          snv  1866318                benign   \n",
       "4   15  67191959     C>G|T          snv  2278670                benign   \n",
       "\n",
       "                 validation_status  \\\n",
       "0  by-frequency;by-alfa;by-cluster   \n",
       "1  by-frequency;by-alfa;by-cluster   \n",
       "2  by-frequency;by-alfa;by-cluster   \n",
       "3  by-frequency;by-alfa;by-cluster   \n",
       "4  by-frequency;by-alfa;by-cluster   \n",
       "\n",
       "                                      function_class                gene  \\\n",
       "0   genic_upstream_transcript_variant;intron_variant               SMAD3   \n",
       "1  5_prime_UTR_variant;genic_upstream_transcript_...  SMAD3;LOC102723493   \n",
       "2  5_prime_UTR_variant;coding_sequence_variant;sy...               SMAD3   \n",
       "3  genic_upstream_transcript_variant;intron_varia...               SMAD3   \n",
       "4                                3_prime_UTR_variant               SMAD3   \n",
       "\n",
       "                                           frequency  \n",
       "0  T:0.034345:172:1000Genomes|T:0.000259:1:ALSPAC...  \n",
       "1  A:0.223642:1120:1000Genomes|A:0.233783:901:ALS...  \n",
       "2  A:0.165935:831:1000Genomes|A:0.111572:430:ALSP...  \n",
       "3  T:0.034345:172:1000Genomes|T:0.000259:1:ALSPAC...  \n",
       "4  T:0.252796:1266:1000Genomes|T:0.228075:879:ALS...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps = pd.read_csv(\"../data/snps.csv\")\n",
    "snps.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SMAD3', 'SMAD3;LOC102723493', 'gene'], dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps[\"gene\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['15', '#chr'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps[\"#chr\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Even though it was quite obvious, after thses checks we can safely get rid of `gene` and `#chr` columns, since al the regions we're looking at are from **Smad3** gene that is located in the chromose **#15**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = snps.drop(columns=[\"gene\", \"#chr\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>variation</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>validation_status</th>\n",
       "      <th>function_class</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>33601</td>\n",
       "      <td>33601</td>\n",
       "      <td>33601</td>\n",
       "      <td>33601</td>\n",
       "      <td>1108</td>\n",
       "      <td>29768</td>\n",
       "      <td>33600</td>\n",
       "      <td>29061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>30903</td>\n",
       "      <td>1080</td>\n",
       "      <td>6</td>\n",
       "      <td>31983</td>\n",
       "      <td>30</td>\n",
       "      <td>6</td>\n",
       "      <td>208</td>\n",
       "      <td>7896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>pos</td>\n",
       "      <td>C&gt;T</td>\n",
       "      <td>snv</td>\n",
       "      <td>snp_id</td>\n",
       "      <td>clinical_significance</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>genic_upstream_transcript_variant;intron_variant</td>\n",
       "      <td>T:0.000008:1:TOPMED|T:0.:0:ALFA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>658</td>\n",
       "      <td>5656</td>\n",
       "      <td>29443</td>\n",
       "      <td>658</td>\n",
       "      <td>658</td>\n",
       "      <td>13784</td>\n",
       "      <td>11572</td>\n",
       "      <td>1938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          pos variation variant_type  snp_id  clinical_significance  \\\n",
       "count   33601     33601        33601   33601                   1108   \n",
       "unique  30903      1080            6   31983                     30   \n",
       "top       pos       C>T          snv  snp_id  clinical_significance   \n",
       "freq      658      5656        29443     658                    658   \n",
       "\n",
       "                      validation_status  \\\n",
       "count                             29768   \n",
       "unique                                6   \n",
       "top     by-frequency;by-alfa;by-cluster   \n",
       "freq                              13784   \n",
       "\n",
       "                                          function_class  \\\n",
       "count                                              33600   \n",
       "unique                                               208   \n",
       "top     genic_upstream_transcript_variant;intron_variant   \n",
       "freq                                               11572   \n",
       "\n",
       "                              frequency  \n",
       "count                             29061  \n",
       "unique                             7896  \n",
       "top     T:0.000008:1:TOPMED|T:0.:0:ALFA  \n",
       "freq                               1938  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's check the dataset for data types and look for `Nan` value to see if something needs some convertation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos                      object\n",
       "variation                object\n",
       "variant_type             object\n",
       "snp_id                   object\n",
       "clinical_significance    object\n",
       "validation_status        object\n",
       "function_class           object\n",
       "frequency                object\n",
       "dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps.loc[snps[\"pos\"] == \"pos\", \"pos\"] = \"-1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps[\"pos\"] = snps[\"pos\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos                       int64\n",
       "variation                object\n",
       "variant_type             object\n",
       "snp_id                   object\n",
       "clinical_significance    object\n",
       "validation_status        object\n",
       "function_class           object\n",
       "frequency                object\n",
       "dtype: object"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos                          0\n",
       "variation                    0\n",
       "variant_type                 0\n",
       "snp_id                       0\n",
       "clinical_significance    32493\n",
       "validation_status         3833\n",
       "function_class               1\n",
       "frequency                 4540\n",
       "dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One can see that we fixed the data type for the `pos` column and foud some `Nan` values in other column, howeverm they seem insignificant for now, so we won't bother about them right now."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AREs data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Motif-family</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Type</th>\n",
       "      <th>Mean Acc Short</th>\n",
       "      <th>Mean Acc Long</th>\n",
       "      <th>TTP</th>\n",
       "      <th>HuR</th>\n",
       "      <th>Auf1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>67190739</td>\n",
       "      <td>67190744</td>\n",
       "      <td>ATTTA</td>\n",
       "      <td>ATTTA</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166949</td>\n",
       "      <td>ENST00000327367|ENST00000558763|ENST00000560402</td>\n",
       "      <td>Exon^3UTR|Exon|Intron</td>\n",
       "      <td>protein_coding|retained_intron|retained_intron</td>\n",
       "      <td>0.5544</td>\n",
       "      <td>0.5339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>67190846</td>\n",
       "      <td>67190851</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>WTTTW</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166949</td>\n",
       "      <td>ENST00000327367|ENST00000558763|ENST00000560402</td>\n",
       "      <td>Exon^3UTR|Exon|Intron</td>\n",
       "      <td>protein_coding|retained_intron|retained_intron</td>\n",
       "      <td>0.5955</td>\n",
       "      <td>0.5274</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr15</td>\n",
       "      <td>67190846</td>\n",
       "      <td>67190853</td>\n",
       "      <td>TTTTTTT</td>\n",
       "      <td>WWTTTWW</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166949</td>\n",
       "      <td>ENST00000327367|ENST00000558763|ENST00000560402</td>\n",
       "      <td>Exon^3UTR|Exon|Intron</td>\n",
       "      <td>protein_coding|retained_intron|retained_intron</td>\n",
       "      <td>0.5974</td>\n",
       "      <td>0.5376</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr15</td>\n",
       "      <td>67190847</td>\n",
       "      <td>67190852</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>WTTTW</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166949</td>\n",
       "      <td>ENST00000327367|ENST00000558763|ENST00000560402</td>\n",
       "      <td>Exon^3UTR|Exon|Intron</td>\n",
       "      <td>protein_coding|retained_intron|retained_intron</td>\n",
       "      <td>0.6029</td>\n",
       "      <td>0.5377</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr15</td>\n",
       "      <td>67190848</td>\n",
       "      <td>67190853</td>\n",
       "      <td>TTTTT</td>\n",
       "      <td>WTTTW</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000166949</td>\n",
       "      <td>ENST00000327367|ENST00000558763|ENST00000560402</td>\n",
       "      <td>Exon^3UTR|Exon|Intron</td>\n",
       "      <td>protein_coding|retained_intron|retained_intron</td>\n",
       "      <td>0.6031</td>\n",
       "      <td>0.5461</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start       End    Motif Motif-family Strand  \\\n",
       "0      chr15  67190739  67190744    ATTTA        ATTTA      +   \n",
       "1      chr15  67190846  67190851    TTTTT        WTTTW      +   \n",
       "2      chr15  67190846  67190853  TTTTTTT      WWTTTWW      +   \n",
       "3      chr15  67190847  67190852    TTTTT        WTTTW      +   \n",
       "4      chr15  67190848  67190853    TTTTT        WTTTW      +   \n",
       "\n",
       "              Gene                                       Transcript  \\\n",
       "0  ENSG00000166949  ENST00000327367|ENST00000558763|ENST00000560402   \n",
       "1  ENSG00000166949  ENST00000327367|ENST00000558763|ENST00000560402   \n",
       "2  ENSG00000166949  ENST00000327367|ENST00000558763|ENST00000560402   \n",
       "3  ENSG00000166949  ENST00000327367|ENST00000558763|ENST00000560402   \n",
       "4  ENSG00000166949  ENST00000327367|ENST00000558763|ENST00000560402   \n",
       "\n",
       "              Annotation                                            Type  \\\n",
       "0  Exon^3UTR|Exon|Intron  protein_coding|retained_intron|retained_intron   \n",
       "1  Exon^3UTR|Exon|Intron  protein_coding|retained_intron|retained_intron   \n",
       "2  Exon^3UTR|Exon|Intron  protein_coding|retained_intron|retained_intron   \n",
       "3  Exon^3UTR|Exon|Intron  protein_coding|retained_intron|retained_intron   \n",
       "4  Exon^3UTR|Exon|Intron  protein_coding|retained_intron|retained_intron   \n",
       "\n",
       "   Mean Acc Short  Mean Acc Long  TTP  HuR  Auf1  \n",
       "0          0.5544         0.5339  NaN  NaN   NaN  \n",
       "1          0.5955         0.5274  NaN  NaN   NaN  \n",
       "2          0.5974         0.5376  NaN  NaN   NaN  \n",
       "3          0.6029         0.5377  NaN  NaN   NaN  \n",
       "4          0.6031         0.5461  NaN  NaN   NaN  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ares = pd.read_csv(\"../data/ARESite2_motif.csv\")\n",
    "ares.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr15'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ares[\"Chromosome\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "ares = ares.drop(columns=[\"Chromosome\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Mean Acc Short</th>\n",
       "      <th>Mean Acc Long</th>\n",
       "      <th>TTP</th>\n",
       "      <th>HuR</th>\n",
       "      <th>Auf1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.830000e+02</td>\n",
       "      <td>1.830000e+02</td>\n",
       "      <td>183.000000</td>\n",
       "      <td>183.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>6.719369e+07</td>\n",
       "      <td>6.719370e+07</td>\n",
       "      <td>0.570546</td>\n",
       "      <td>0.445738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>9.827207e+02</td>\n",
       "      <td>9.832639e+02</td>\n",
       "      <td>0.099612</td>\n",
       "      <td>0.065839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>6.719074e+07</td>\n",
       "      <td>6.719074e+07</td>\n",
       "      <td>0.386700</td>\n",
       "      <td>0.352100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>6.719369e+07</td>\n",
       "      <td>6.719370e+07</td>\n",
       "      <td>0.494200</td>\n",
       "      <td>0.378950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>6.719372e+07</td>\n",
       "      <td>6.719373e+07</td>\n",
       "      <td>0.566100</td>\n",
       "      <td>0.452500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>6.719450e+07</td>\n",
       "      <td>6.719451e+07</td>\n",
       "      <td>0.667150</td>\n",
       "      <td>0.484950</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>6.719515e+07</td>\n",
       "      <td>6.719516e+07</td>\n",
       "      <td>0.818700</td>\n",
       "      <td>0.752200</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Start           End  Mean Acc Short  Mean Acc Long  TTP  HuR  \\\n",
       "count  1.830000e+02  1.830000e+02      183.000000     183.000000  0.0  0.0   \n",
       "mean   6.719369e+07  6.719370e+07        0.570546       0.445738  NaN  NaN   \n",
       "std    9.827207e+02  9.832639e+02        0.099612       0.065839  NaN  NaN   \n",
       "min    6.719074e+07  6.719074e+07        0.386700       0.352100  NaN  NaN   \n",
       "25%    6.719369e+07  6.719370e+07        0.494200       0.378950  NaN  NaN   \n",
       "50%    6.719372e+07  6.719373e+07        0.566100       0.452500  NaN  NaN   \n",
       "75%    6.719450e+07  6.719451e+07        0.667150       0.484950  NaN  NaN   \n",
       "max    6.719515e+07  6.719516e+07        0.818700       0.752200  NaN  NaN   \n",
       "\n",
       "       Auf1  \n",
       "count   0.0  \n",
       "mean    NaN  \n",
       "std     NaN  \n",
       "min     NaN  \n",
       "25%     NaN  \n",
       "50%     NaN  \n",
       "75%     NaN  \n",
       "max     NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ares.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Start               int64\n",
       "End                 int64\n",
       "Motif              object\n",
       "Motif-family       object\n",
       "Strand             object\n",
       "Gene               object\n",
       "Transcript         object\n",
       "Annotation         object\n",
       "Type               object\n",
       "Mean Acc Short    float64\n",
       "Mean Acc Long     float64\n",
       "TTP               float64\n",
       "HuR               float64\n",
       "Auf1              float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ares.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Start               0\n",
       "End                 0\n",
       "Motif               0\n",
       "Motif-family        0\n",
       "Strand              0\n",
       "Gene                0\n",
       "Transcript          0\n",
       "Annotation          0\n",
       "Type                0\n",
       "Mean Acc Short      0\n",
       "Mean Acc Long       0\n",
       "TTP               183\n",
       "HuR               183\n",
       "Auf1              183\n",
       "dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ares.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All the data types are okay and so are the `Nan` values (I mean their absence), so let's move on to comparing SNPs and AREs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Polymorphism in AREs search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_variation(variation):\n",
    "    divider = variation.find('>')\n",
    "    starting_allele = variation[:divider - 1]\n",
    "    polymorphed_alleles = tuple([ch for ch in variation[divider:].strip().split('|')])\n",
    "    return starting_allele, polymorphed_alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8ae4da9580e343b0bf23a962f6123d0b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "|          | 0/? [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyError",
     "evalue": "'motif'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/usr/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2894\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2895\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2896\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'motif'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-38-514818e19580>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mpos\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m                 \u001b[0ms_allel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpoly_allels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_variation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msnp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"variation\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m                 \u001b[0mmotif\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mare\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"motif\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0ms_allel\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0mmotif\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mare\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Start\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mpos\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mare\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Start\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mpos\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms_allel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0ms_allel\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    880\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    881\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mkey_is_scalar\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 882\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    883\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    884\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_hashable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36m_get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m    987\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    988\u001b[0m         \u001b[0;31m# Similar to Index.get_value, but we do not fall back to positional\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 989\u001b[0;31m         \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    990\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_values_for_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    991\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   2895\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcasted_key\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2896\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2897\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2898\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2899\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtolerance\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'motif'"
     ]
    }
   ],
   "source": [
    "are_with_snp = list()\n",
    "\n",
    "for _, are in tqdm(ares.iterrows()):\n",
    "    for _, snp in snps.iterrows(): \n",
    "        if are[\"Start\"] <= int(snp[\"pos\"]) <= are[\"End\"]:\n",
    "            pos = int(snp[\"pos\"])\n",
    "            if pos > 0:\n",
    "                s_allel, poly_allels = parse_variation(snp[\"variation\"])\n",
    "                motif = are[\"Motif\"]\n",
    "                if s_allel != '-':\n",
    "                    if motif[are[\"Start\"] - pos:are[\"Start\"] - pos + len(s_allel)] != s_allel:\n",
    "                        print(\"! allele doesn't coinside with motif sequence\")\n",
    "                        continue\n",
    "                alleles = s_allel\n",
    "                for allele in poly_allels:\n",
    "                    alleles += \"/\" + allele\n",
    "                motif = motif[:are[\"Start\"] - pos] + alleles + motif[are[\"Start\"] - pos + len(s_allelallel):]\n",
    "\n",
    "                are_with_snp.append((motif, are[\"Start\"], are[\"End\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
